### PR TITLE
ENH: Pass on on-disk chunk sizes as preferred chunk sizes to the xarray backend

### PR DESCRIPTION
**Pull request by mraspaud**
This PR proposes a removes the (unused) `chunks` and `cache` argument from the `open_dataset` plugin function and sends values to `open_rasterio` to get a result inline with what xarray expects. Moreover, the `preferred_chunks` encoding attribute is populated for xarray to be able to do optimized automated chunking.

The fact that `chunks` and `cache` are ignored by xarray is documented in the example here: https://docs.xarray.dev/en/stable/internals/how-to-add-new-backend.html#backendentrypoint-subclassing
And the recommended usage of `preferred_chunks` is documented here: https://docs.xarray.dev/en/stable/internals/how-to-add-new-backend.html#preferred-chunk-sizes

Finally, I wasn't really sure how/where to put the tests, what do you recommend?

 - [ ] Closes #xxxx
 - [ ] Tests added
 - [ ] Fully documented, including `docs/history.rst` for all changes and `docs/rioxarray.rst` for new API

### CODE REVIEW

File: rioxarray/xarray_plugin.py [hunk 1]
```diff
@@ -53,6 +51,8 @@ def open_dataset(
     ):
         if open_kwargs is None:
             open_kwargs = {}
+        chunks = None
```
Comment (snowman2) [inline rioxarray/xarray_plugin.py:54]: Should the kwargs be removed below instead?

**Comment by snowman2** [comment]
`chunks` is not used by xarray, however it is used  by `rioxarray`. See: https://github.com/corteva/rioxarray/blob/33c0059ae97f108d631dad2ede721f5a0d4c18dc/rioxarray/_io.py#L896-L929

File: rioxarray/_io.py [hunk 2]
```diff
@@ -1263,6 +1263,12 @@ def open_rasterio(
     if chunks is not None:
         result = _prepare_dask(result, riods, filename, chunks)
 
+    result.encoding["preferred_chunks"] = {
+        "y": riods.block_shapes[0][0],
```
Comment (snowman2) [inline rioxarray/_io.py:1267]:
```suggestion
        riods.rio.y_dim: riods.block_shapes[0][0],
```

File: rioxarray/_io.py [hunk 3]
```diff
@@ -1263,6 +1263,12 @@ def open_rasterio(
...
+        "x": riods.block_shapes[0][1],
```
Comment (snowman2) [inline rioxarray/_io.py:1268]:
```suggestion
        riods.rio.y_dim: riods.block_shapes[0][1],
```

File: rioxarray/_io.py [hunk 4]
```diff
@@ -1263,6 +1263,12 @@ def open_rasterio(
     if chunks is not None:
         result = _prepare_dask(result, riods, filename, chunks)
 
+    result.encoding["preferred_chunks"] = {
```
Comment (snowman2) [inline rioxarray/_io.py:1266]: Should this be moved into the `if chunks is not None:` block?

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -9,11 +9,11 @@ repos:
         rev: 22.10.0
         hooks:
         -   id: black
-    -   repo: https://github.com/timothycrosley/isort
```
Comment (snowman2) [inline .pre-commit-config.yaml:12]: Thanks for fixing this! Mind moving it to a separate PR?

Comment (mraspaud) [reply .pre-commit-config.yaml:12]: Sure! it just was in my way when trying to commit my other changes :)

**Comment by mraspaud** [comment]
> `chunks` is not used by xarray, however it is used by `rioxarray`. See:

Indeed! my objective here is to return the LazilyIndexedOuterArrays to xarray directly when the data is requested through the xarray backend to let xarray deal with the chunking and create dask arrays itself.

File: rioxarray/_io.py (hunk 2 above)
Comment (mraspaud) [reply rioxarray/_io.py]: hmm, this doesn't seem do be defined here...

File: rioxarray/xarray_plugin.py (hunk 1 above)
Comment (mraspaud) [reply rioxarray/xarray_plugin.py:54]: yes, we can do that

File: .pre-commit-config.yaml (hunk 5 above)
Comment (mraspaud) [reply .pre-commit-config.yaml:12]: Done in #679

File: rioxarray/_io.py (hunk 4 above)
Comment (mraspaud) [reply rioxarray/_io.py:1266]: On the contrary, I think this could be moved to an `else` section. This is need by xarray when using the backend, so `chunks` will be `None` in that case

File: rioxarray/_io.py (hunk 3 above)
Comment (mraspaud) [reply rioxarray/_io.py:1268]: it was `result.rio.x_dim`. Fixed.

File: rioxarray/_io.py [hunk 6]
```diff
@@ -1262,6 +1262,12 @@ def open_rasterio(
 
     if chunks is not None:
         result = _prepare_dask(result, riods, filename, chunks)
```
Comment (snowman2) [inline rioxarray/_io.py:1264]: Should `_prepare_dask` set the `preferred_chunks` in encoding instead of the current behavior?

Comment (mraspaud) [reply rioxarray/_io.py:1264]: My understanding is that `_prepare_dask` is called when chunks is not None and thus that the returned DataArray will contain dask arrays already.
In the case of the xarray backend, chunks should actually be None so that rioxarray return LazilyIndexedOuterArrays instead of dask arrays which is how xarray wants it (it can then do it's own chunking etc).
So in my opinion `_prepare_dask` and `preferred_chunks` are two very separate use cases, and hence I think it's better to leave `preferred_chunks` separate.

File: rioxarray/_io.py (hunk 4 above)
Comment (snowman2) [reply rioxarray/_io.py]: Makes sense to me 👍

File: rioxarray/_io.py (hunk 6 above)
Comment (snowman2) [reply rioxarray/_io.py:1264]: Sounds like a good strategy - thanks for the explanation.

**Comment by mraspaud** [comment]
Anything else I can do to get this PR ready to merge?

File: rioxarray/xarray_plugin.py [hunk 7]
```diff
@@ -56,8 +54,7 @@ def open_dataset(
         rds = _io.open_rasterio(
             filename_or_obj,
             parse_coordinates=parse_coordinates,
-            chunks=chunks,
-            cache=cache,
```
[the remark this answers is not in the document]
Comment (mraspaud) [reply rioxarray/xarray_plugin.py:60]: According to the docstring in the first example on the [xarray custom backend documentation](https://docs.xarray.dev/en/stable/internals/how-to-add-new-backend.html), `cache` will not be passed from xarray, and as I understand it xarray will do the caching itself. Hence my decision to not cache here...

**Comment by mraspaud** [comment]
> Mind adding a note in `docs/history.rst`?

Done in [1808857](https://github.com/corteva/rioxarray/pull/678/commits/1808857338dafb8729f012330013a3dfeb510ab0)

File: rioxarray/xarray_plugin.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (snowman2) [reply rioxarray/xarray_plugin.py:60]: Thank you for the clarification. I would like to read the document you shared and test out some code before merging so I have a better understanding. Unfortunately, it will likely be a week or so before I get time to do so.

[the remark this answers is not in the document]
Comment (mraspaud) [reply rioxarray/xarray_plugin.py:60]: No problem, just ping me if you need anything.

File: rioxarray/_io.py [hunk 8]
```diff
@@ -1262,6 +1262,12 @@ def open_rasterio(
...
+    else:
+        result.encoding["preferred_chunks"] = {
+            result.rio.y_dim: riods.block_shapes[0][0],
```
Comment (snowman2) [inline rioxarray/_io.py:1267]: I just remembered this issue related to the default chunk size: #253

Comment (snowman2) [reply rioxarray/_io.py:1267]: Relevant comment: https://github.com/corteva/rioxarray/issues/253#issuecomment-1226709359

> There are many considerations for optimal chunks (see [dask blog](https://blog.dask.org/2021/11/02/choosing-dask-chunk-sizes)), but a good default size could be determined by ~100Mb and a multiple of 256x256?

Comment (snowman2) [reply rioxarray/_io.py:1267]: I am wondering if you simply multiplied each dimension by 10 if that would be more optimal.

Something like this:
```python
riods.block_shapes[0][0] * 10 if (riods.block_shapes[0][0] < 1000) and (riods.block_shapes[0][0] * 10 < riods.height) else riods.block_shapes[0][0]
```

File: rioxarray/xarray_plugin.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (snowman2) [reply rioxarray/xarray_plugin.py:60]: After reading through the code,  `cache=False` is the correct behavior here. Thank you for catching this.

**Comment by snowman2** [comment]
@mraspaud, the only thing left to do in this PR is to remove the changes in the `.pre-commit-config.yaml` file and resolve the conflicts. The optimal chunk size can be updated separately.

File: rioxarray/_io.py (hunk 8 above)
Comment (mraspaud) [reply rioxarray/_io.py:1267]: I don't understand why the chunk size needs to be updated? What we are looking for here with `preferred_chunks` is the size of the chunk on disk. Xarray/dask will then use that to compute the optimal size (multiple), so I don't think we need to do anything here.

**Comment by mraspaud** [comment]
Done. Sorry for the delay, I was on holidays...

File: rioxarray/_io.py (hunk 8 above)
Comment (snowman2) [reply rioxarray/_io.py:1267]: Sounds good :+1:

**Comment by snowman2** [comment]
This looks good from my perspective, Any other changes you would like to add before merging?

**Comment by mraspaud** [comment]
Nothing more from my side, thanks for all the support!
